### ISSUE TITLE
[no-useless-constructor] Skipping constructors with explicit member accessibility

### ISSUE THREAD

**Issue by rhalaly** [typescript-eslint/typescript-eslint#948]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-useless-constructor": "error",
  }
}
```

```TS
public constructor(a: int) {
   super(a)
}
```

**Expected Result**
Error useless constructor
**Actual Result**
No error
**Additional Info**
Working fine without the `public` in the decleration

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.1.0` |
| `@typescript-eslint/parser`               | `2.1.0` |
| `TypeScript`                              | `3.5.3` |
| `ESLint`                                  | `6.3.0` |
| `node`                                    | `8.11.3` |
| `npm`                                     | `6.11.2` |

**Comment by bradzacher** [comment]
How come you're changing the accessibility of the constructor, that seems like you're breaking encapsulation?

This is a bug, but I'm not sure if it's something we can fix easily. Fixing it would require type information, which is a breaking change in the rule.

In the interim, I'd suggest that you use an eslint disable comment for this one case.

**Comment by rhalaly** [comment]
Not sure why it is braking change, just ignore the accessibility.

I just noticed this bug when I used the rule "explicit member accessibility" without the constructor no-public override.

I'm not sure that it is a common bug, but this is a bug.

**Comment by bradzacher** [comment]
```ts
class A { public constructor() {} }
class B extends A ( public constructor() { super() } }
class C extends A ( private constructor() { super() } }
```

In the above example, `B` clearly has a useless constructor - it has an explicit accessibility.
`C` is not a useless constructor, because it changes the accessibility.

That's exactly the point - the explicit member accessibility rule prompts people to add an accessibility to constructors. But in the general case, chances are they're going to use the same accessibility as the parent constructor, which is a useless constructor.

So in order to properly handle this case, we cannot just do `constructorHasAccessibility ? valid : invalid`, we instead need to do `parentConstructorHasDifferentAccessibility ? valid : invalid`. To do this, it requires type information - there is no guarantee that the parent class exists in your current file, which means we cannot detect it via just checking the AST.

Bringing in type information is considered a breaking change, as users will have to change their config in order to keep the rule working.

I'm not saying we won't look to fix this - it's a valid case we should handle, but it won't necessarily happen "soon".

**Comment by rhalaly** [comment]
Ho, I didn't think about that so deeply. You are right.

Thanks for the explanation!

**Comment by cowwoc** [comment]
@bradzacher To clarify why users would want to change the accessibility of classes:

```
abstract class Parent { protected constructor() {} }
class Child extends Parent { public constructor() { super() } }
```

So we're not breaking accessibility. We're exposing certain sub-classes to users while keeping others private.

**Comment by JoshuaKGoldberg** [comment]
From https://github.com/typescript-eslint/typescript-eslint/pull/6045#issuecomment-1325876143:

> Yeah, you're right. The lack of activity from more than 3-4 people -only 2 👍s- is good evidence that this isn't a particularly necessary change.
> 
> +1 to what you pointed out in the issue that this is a very odd thing to do. OOP with extended classes messing with blank constructors is such a rare case in most modern JS codebases.
> 
> I'll close this PR & issue for now, but if someone has a real strong reason why we should reopen, please feel free to comment back and/or fork the rule in your own codebase!

**Comment by cowwoc** [comment]
From https://github.com/typescript-eslint/typescript-eslint/pull/6045#issuecomment-1326516643:

> Use-case:
> 
> ```
> abstract class Parent { protected constructor() {} }
> class Child extends Parent { public constructor() { super() } }
> ```
> 
> Users shouldn't be able to invoke the `Parent` constructor, but invoking the `Child` constructor is fair game. Does this help?

**Comment by JoshuaKGoldberg** [comment]
That is a theoretically valid case & argument for why we should do this, yes! But it feels like a very odd one that isn't hit by most users. Why are you using _inheritance_ instead of _composition_ - given that [Composition over Inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) is a core principle of modern OOP? What _common_, _good_, _important-to-use_ architecture is at play here?

Adding type checking information to the rule makes it slower, restricts some users from being able to use it, and increases the complexity of the rule - making it harder to change later down the line. What I'm looking for is evidence that there are _good and non-rare_ architectural patterns that this bug is preventing.

So, while your use case is totally valid and evidence in favor of this issue, we haven't yet seen enough evidence to convince us the downsides of making the rule type-informed aren't more significant.

FWIW my PR at #6045 shows a working version of the rule that does take type information into account. You can always publish that as a standalone ESLint plugin and use it if you really want! Or, [`// eslint-disable-next-line`](https://eslint.org/docs/latest/user-guide/configuring/rules#disabling-rules) on these specific cases.

**Comment by cowwoc** [comment]
> given that [Composition over Inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) is a core principle of modern OOP? What common, good, important-to-use architecture is at play here?

That's a rather subjective determination. I routinely make use of _inheritance_ and _composition_ in my daily work, choosing whichever one is appropriate for the circumstance. Modern popular programming languages continue to use inheritance heavily and introduce new APIs that benefit from their use. I'll give you a concrete example introduced by Java no more than 2 months ago (as part of the JDK 19 release): `java.lang.Thread` and `java.util.concurrent.ThreadFactory` were extended to support the concept of _Virtual Threads_. Had the API made use of composition, it would have been impossible to introduce such a feature in a transparent manner. Concretely, users are able to construct an instance of `java.util.concurrent.Executor` backed by virtual threads into libraries that were written prior to the introduction of virtual threads, and those libraries make use of virtual threads without having to change a line of code.

Consequently, I argue that there are plenty of legitimate use-cases for making use of _inheritance_ and such a basic/core feature of the Typescript language should be supported by ESLint.

**Comment by Josh-Cena** [comment]
> ```js
> abstract class Parent { protected constructor() {} }
> class Child extends Parent { public constructor() { super() } }
> ```

I'm slightly confused by this example—I suppose to meant to make `Parent` non-abstract? Otherwise `Parent` is not constructible anyway, and you can remove the visibility markers from both places.

**Comment by JoshuaKGoldberg** [comment]
@cowwoc that's Java, we're talking about JavaScript. The two are not the same language.

😛

I'm kidding - but, what we're asking for is real world counterexamples and architecture patterns in JavaScript. Many JS/TS codebases do not use composition or even many classes at all these days, so we want to see more evidence in favor of adding this bugfix.

**Comment by cowwoc** [comment]
@JoshuaKGoldberg Fair enough.

@Josh-Cena 

> > ```ts
> > abstract class Parent { protected constructor() {} }
> > class Child extends Parent { public constructor() { super() } }
> > ```
> 
> I'm slightly confused by this example—I suppose you meant to make `Parent` non-abstract? Otherwise `Parent` is not constructible anyway, and you can remove the visibility markers from both places.

You're right. This is a poor example. I'll leave this alone until the next time I run across this issue. Hopefully I'll have a more concrete counter-example by then :) Thanks guys.
